sdk/core/src/headers/mod.rs

```diff
@@ -178,13 +193,14 @@ impl HeaderName {
 
 impl From<&'static str> for HeaderName {
     fn from(s: &'static str) -> Self {
+        assert_eq!(s.to_lowercase(), s, "header names must be lower-case");
```

Comment:
This probably should be inside `HeaderName::from_static` or even better in a private constructor that takes a `Cow<'static, str>` and constructs the `HeaderName` which everything uses instead of constructing the `HeaderName` directly.

Reply:
While I'd prefer this to live in from_static, I did not want to break the `const` attribute, which @cataggar was invested in adding, none of the methods I can come up with are const.  (for loops, chars iterator, and to_lowercase)

I added a private from_cow method which is used by the From impls that does this.